modules/javafx.graphics/src/main/java/javafx/scene/Node.java

```diff
@@ -1404,6 +1405,87 @@ public String getName() {
         return visible;
     }
 
+    /**
+     * Whether or not this {@code Node} is showing (that is, it is part of an
+     * open Window on the user's system). The Window might be "showing", yet the user might not
+     * be able to see it due to the Window being rendered behind another Window
+     * or due to the Window being positioned off the monitor.
+     *
+     * <p>Note that the {@code Node} does not need to be visible for this property
+     * to be {@code true}.
+     *
+     * @defaultValue false
+     * @since 20
+     */
+    private ReadOnlyBooleanProperty shown;
```

Comment:
I would recommend against adding this property, as it adds a pointer to every Node regardless of whether it's needed or not.

Another reason not to add is that, at least in my experience, there is rarely a need to base the logic on whether a particular Node is shown or not - instead, the logic to disconnect (one or more) listeners should be based on a Window/Dialog or a container pane.  And for that I think a better solution might be - sorry for self plug - something like ListenerHelper https://github.com/openjdk/jfx/pull/908

And another thing - Window and Dialog are not Nodes, but they do have `showing` property.

I would suggest either remove the changes pertaining to Node, or perhaps provide a utility method to create the property and store it in Node.getProperties() or Window.getProperties() (sadly, there is no Dialog.getProperties())

Reply:
I agree with your first point, but I think limiting any changes to `Node` because it would add another field is not a good enough argument.  It would depend on its merits.

The second point, there is much more to the `shownProperty` then you might think at first glance. In combination with `when`, it offers a standard way to manage life cycles for controls.  It is a very natural fit to tie the lifecycle of bindings and listeners to whether a control is actually currently shown.  You may think that a Listener helper class would be just as flexible, but it isn't.  The helper is not available everywhere (unless you propose to integrate it into Node), which means when constructing complicated UI's which may have smaller groups of controls that are created by other classes, methods or factories, that you may have to pass along an instance of your helper to all these to make sure they register their bindings and listeners with it (or perhaps to link multiple Listener helper classes together in some kind of hierarchy, that matches the control groupings that might have separate life cycles).

The `shownProperty` does not have this problem. Since you can tie the lifecycle to your own small piece of the UI (ie, a factory that creates a `BorderPane` with several controls, can tie all these to the lifecycle of the `BorderPane`).  When this `BorderPane` is just a piece of UI used as part of a larger whole, its lifecycle still works as expected -- if it is not shown, its listeners disappear.  If the whole window is destroyed, then the same happens.  If however the Pane next to it is destroyed, only their listeners disappear.  With a helper system, you'd need to coordinate this somehow, perhaps having that `BorderPane` return its helper, so the large UI can tie it to its helper, so when the main helper is called to unsubscribe, that all its child helpers are also unsubscribed...

Another advantage is that a listener does not have to start working right away.  When setting up a new pane of controls, many subscriptions may need to be made.  When using `when(control::shownProperty)` these listeners don't immediately become active (potentially while still constructing and initializing your control).  Only when the control is fully constructed, and added to an active Scene and Window do all the listeners become active.  This can be important when listening to external properties that change often and may even change during construction.

In ReactFX it the `shownProperty` was deemed important enough to go one step further and even create a separate method on what would be `ObservableValue` called `conditionOnShowing`.  

In short, `when` and the `shownProperty` is basically used every time you create a binding or add a listener between two properties that have a different lifecycle, like between a model and a control.  This completely removes the responsibility of having to track these bindings or listeners separately; it becomes automatic. Subscribing all at once when it becomes first shown becomes automatic (and you immediately get the first value), unsubscribing when it goes invisible is automatic, and resubscribing if a control becomes shown again is completely transparent.

In my projects, I've completely eliminated memory leaks without a single call to `removeListener` or `unbind` anywhere. 
There are no explicit weak listeners (and not just because they're incredibly cumbersome to use), nor is there anywhere where subscriptions or bindings are being tracked (aside from the `when` construct).  I'm very picky about memory leaks, so I have a Scene scanner, that maintains weak references to all Nodes at all times, and warns me if any are not getting GC'd after not being part of a Scene for a while.  In 99% of the cases, I forgot a `when`, where in plain JavaFX you'd be forced to use a weak listener.

The simple rule to follow is, whenever you bind or listen to or on a property, if they have different lifecycles, then use a `when`, otherwise bind or listen directly.

Some examples of how it is used:

Below is a very common pattern when binding any property that has a different lifecycle than its target, this is one of many:

    longLivedModel.selectedItem
         .when(this::shownProperty)   // this = BorderPane, and the code is part of its constructor)
         .addListener((obs, old, current) -> {
              // code that starts a transition from the previous item to the next item
         }
    );

Below some binding examples where several properties are bound to long lived versions:

    ObservableValue<Boolean> shown = pane.shownProperty();

    pane.model.work.bind(presentation.root.when(shown));
    pane.model.missingFraction.bind(presentation.missingFraction.when(shown));
    pane.model.watchedFraction.bind(presentation.watchedFraction.when(shown));

Below a "bidirectional binding" for a `Slider` control, where one direction is long lived.  The binding vanishes on its own when the UI that is using it is closed -- this is part of a UI generator, so this code has no idea where the UI might end up being used or for how long, but it can still manage the lifecycle easily:

    longLivedValue.conditionOn(slider::shownProperty).map(toNumber).subscribe(slider.valueProperty()::setValue);
    slider.valueProperty().map(fromNumber).subscribe(value::setValue);

Another example to make sure a `Timeline` is stopped/started automatically based on visibility (Timeline's will keep references around if not stopped):

    center.shownProperty()  // center here is a Pane
      .addListener((obs, old, visible) -> {
        if(visible) timeline.playFromStart();
        else timeline.stop();
      });

So, I think it certainly has some merit to include it :)

> And another thing - Window and Dialog are not Nodes, but they do have showing property.

I'm unsure what you are saying here.  I don't think we need to tie anything directly to a Window or Dialog, only to controls they might contain.  Also the `shownProperty` is not useful for properties not associated with controls; depending on what the properties are used for people may have to create their own `active` property or something else that identifies the lifecycle of those properties.  For control properties though, their lifecycle is quite clear and I think it makes great sense to tie to whether they're actually visible or not.

Reply:
What I meant is you can add it in Node.getProperties() and not to add a pointer to every Node everywhere.  If it's used, the actual properties will be created in a few containers.

Reply:
```
private static final Object KEY = new Object();

public final ReadOnlyBooleanProperty shownProperty() {
  Object x = getProperties().get(KEY);
  if (x instanceof ReadOnlyBooleanProperty p) {
    return p;
  }
  ReadOnlyBooleanProperty p = ... // create
  getProperties().put(KEY, p);
  return p;
}


```

Reply:
Thanks, I understand, what I find odd is that this would be the first property in `Node` to be stored this way.  If this was so important, then why isn't this done for other properties? There seem to be sufficient candidates for this approach to "slim" down `Node` (infrequently used properties like `Subscene`, `id`, `style`, `blendMode`).

There's even a `MiscProperties` objects which is described as "Misc Seldom Used Properties" that is created only when needed with these properties:

        private LazyBoundsProperty boundsInParent;
        private LazyBoundsProperty boundsInLocal;
        private BooleanProperty cache;
        private ObjectProperty<CacheHint> cacheHint;
        private ObjectProperty<Node> clip;
        private ObjectProperty<Cursor> cursor;
        private ObjectProperty<DepthTest> depthTest;
        private BooleanProperty disable;
        private ObjectProperty<Effect> effect;
        private ObjectProperty<InputMethodRequests> inputMethodRequests;
        private BooleanProperty mouseTransparent;
        private DoubleProperty viewOrder;

Let's see what others think before I change this. It does sound like a reasonable approach though.

Reply:
I don't have time to review this as I am busy with JavaOne, so this is just a quick note to say that if there is sufficient benefit to add a `shown` property to the core of JavaFX, then it should be added _as_ a property -- possibly in MiscProperties if we think it is a seldom-used property. I don't like the idea of using `getProperties` for this.

Reply:
> I'd respectfully disagree: in a large application, there might be hundreds (a thousand?) Nodes, and only a handful (10-20?) Nodes with this property (2%), so it's a waste of RAM. One may think it's only 8Kb on a 64 bit machine, but it does add up quickly. Even the access pattern is basically 'get it once to establish the fluent pipeline', with no more calls to the Node accessor (unlike MiscProperties constituents, some of which get accessed rather frequently).

I understand your point, but I don't think that this is sufficient reason to introduce a new way of storing property instances in an ad-hoc manner. If memory consumption is a concern, I would rather analyze the issue in a separate enhancement proposal, think it through in terms of API, and then transition all relevant property instances to the new storage mechanism. Part of that proposal should also be an evaluation of the trade-offs between memory consumption and performance in terms of cache locality and lookup speed.

Reply:
@mstr2 : 
`getProperties()` is a perfectly fine way of storing a property instance, especially considering the fact that this property will be called exactly once - during the fluent pipeline setup.  Of course, it does not apply to other properties that are being called all the time.

A separate ticket for optimizing property storage might be a good idea, I agree.

Reply:
> > `getProperties()` isn't the place for this.
> 
> can you explain why?

This map is used for storing values for the node. As the API note says, layout managers use this to store constraints that be updated by the user. I think that this is already not the best way to do it. The user can also store arbitrary data like markers or other associations. Storing a `Property` is like storing functionality, not data.

In addition, the map is publicly accessible, so the user can clear it. In this case you risk GC of the instance. This is equivalent to making a property field both `public` and not `final`, allowing the user to overwrite the reference - a rather disastrous scenario.

Reply:
Thank you.  You do bring a good point - clearing the map.  There is some mild language warning about clearing the map in javadoc already, but perhaps `Node.getProperties().clear()` should be forbidden (throw an IOE?).  

Also, the devs risk unexpected side effects if they try to use Strings as keys, due to possible collision with keys used by FX internals.

And, of course, the API does not preclude storing properties in `getProperties()`, and FX is already doing it, as illustrated by ButtonBarSkin:136 (perhaps more, I did not perform an exhaustive check).

Reply:
> Thank you. You do bring a good point - clearing the map. There is some mild language warning about clearing the map in javadoc already, but perhaps `Node.getProperties().clear()` should be forbidden (throw an IOE?).

A map can also be cleared with `Map.keySet().clear()`, or by enumerating its keys and removing each mapping one by one.
I don't think we can make the `getProperties()` map mutable, but at the same time safe to store final objects, while still respecting the `Map` contract.

Maybe we could have something like "hidden" keys, which are not enumerable and are never cleared by bulk operations.

Reply:
I think we can conclude here that this is a bit outside the scope of this PR.

I certainly wouldn't be in favor of wrapping the property map to prevent clearing or any such thing; if we want to save space in `Node` this way, a dedicated structure that can't be accessed by users would be the way to go in my opinion.  The space we save should compensate for the pointer to this structure.  

Before we'd even consider that though, I'd like to see where most memory is used in FX so efforts can be focused in areas where it would net the greatest effect.  For example, Properties are far more numerous than nodes.

I noticed the use of the properties map by JavaFX itself -- the API states that this is the case, but I wonder if that's really required.  It might be better to keep these things in `WeakHashMap`, which is intended for this purpose instead of strewing these properties about over the various nodes which are then forced to instantiate the properties map vs just a single `WeakHashMap`.

For example, `ButtonBarSkin` seems to do it to keep track of a variable amount of listeners. It effectively creates the properties map for each button added (assuming the map is usually empty) vs a single `WeakHashMap`; also it does this on an `ObservableList`.  This looks like something that could be a standard pattern (given an observable collection, automatically add/remove a given listener as the collection changes).  ReactFX had `Subscription#dynamic` for this.

It also looks like `ButtonBarSkin` might be buggy, which it wouldn't have been if had used a `WeakHashMap`.  I see no `dispose` method and the listeners added don't use the `register` system, which means it can leave behind listeners that it stores in properties.  A new skin may conclude the listener on some buttons is already present.  Or an old skin may not get garbage collected.

Reply:
FYI, `ButtonBarSkin` does have memory leak issues, which are addressed in https://github.com/openjdk/jfx/pull/921